ntoskrnl/ke/amd64/xstate.c

```diff
@@ -0,0 +1,270 @@
+    XStateConfig->EnabledSupervisorFeatures = SupportedSupervisorMask;
+
+     /* Save the calculated sizes */
+    XStateConfig->Size = NextUserOffset;
+    XStateConfig->AllFeatureSize = NextSupervisorOffset;
+    ASSERT(XStateConfig->AllFeatureSize >= XStateConfig->Size);
+}
+
+/*!
+ *   \brief Validates the provided extended state configuration against the global one
+ *
+ *   \param XStateConfig - Pointer to a XSTATE_CONFIGURATION structure containing the configuration
+ */
+CODE_SEG("INIT")
+static
+VOID
+ValidateXStateConfig(
+    _In_ PXSTATE_CONFIGURATION XState)
+{
+    PXSTATE_CONFIGURATION GlobalXState = &SharedUserData->XState;
+
+    if ((XState->EnabledFeatures != GlobalXState->EnabledFeatures) ||
+        (XState->EnabledSupervisorFeatures != GlobalXState->EnabledSupervisorFeatures) ||
+        (XState->Size != GlobalXState->Size) ||
+        (XState->AllFeatureSize != GlobalXState->AllFeatureSize))
+    {
+        /* Invalid features */
+        KeBugCheck(MULTIPROCESSOR_CONFIGURATION_NOT_SUPPORTED);
+    }
+
+    for (ULONG i = 0; i < MAXIMUM_XSTATE_FEATURES; i++)
+    {
+        if ((XState->Features[i].Size != GlobalXState->Features[i].Size) ||
+            (XState->Features[i].Offset != GlobalXState->Features[i].Offset) ||
+            (XState->AllFeatures[i] != GlobalXState->AllFeatures[i]))
+        {
+            /* Invalid features */
+            KeBugCheck(MULTIPROCESSOR_CONFIGURATION_NOT_SUPPORTED);
+        }
+    }
+}
+
+/*!
+ *   \brief Initializes the extended state configuration for the current processor
+ *
+ *   \param ProcessorNumber - Number of the current processor
+ */
+CODE_SEG("INIT")
+VOID
+NTAPI
+KiInitializeXStateConfiguration(
+    _In_ ULONG ProcessorNumber)
+{
+    if (ProcessorNumber == 0)
+    {
+        /* Processor 0: Retrieve the global configuration */
+        KiGetXStateConfiguration(&SharedUserData->XState);
+        KeXStateLength = SharedUserData->XState.AllFeatureSize;
+    }
+    else
+    {
+        /* Application processor: validate the configuration against the global one */
+        XSTATE_CONFIGURATION XState;
+        KiGetXStateConfiguration(&XState);
+        ValidateXStateConfig(&XState);
+    }
+
+    /* Enable the user mode components in XCR0 */
+    _xsetbv(0, SharedUserData->XState.EnabledFeatures);
+
+    /* Now that we have set everything up, query CPUID again to get the required
+       size based on components enabled in XCR0 */
+    CPUID_EXTENDED_STATE_MAIN_LEAF_REGS ExtStateMain;
+    __cpuidex(ExtStateMain.AsInt32,
+              CPUID_EXTENDED_STATE,
+              CPUID_EXTENDED_STATE_MAIN_LEAF);
+
+    /* Make sure our calculation adds up! We only make sure that our size is
+       large enough, as some VMs always report the full size, independent of
+       what has been enabled in XCR0 */
+    if (ExtStateMain.Ebx > SharedUserData->XState.Size)
+    {
+        KeBugCheck(MULTIPROCESSOR_CONFIGURATION_NOT_SUPPORTED);
```

Comment:
Maybe you could display a bit more of information with this bsod, like the encountered ebx value and the shared user data xstate size ?
(something like what was done in https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/ke/i386/kiinit.c;hb=0ca6002b4fbab4cd1e3c180e5c9cc987dae891d8#l253 )

Reply:
I removed it, it's pointless